### DISCUSSION TITLE
Mimir Write in Python fails with err-mimir-missing-metric-name

### DISCUSSION THREAD

**Discussion by Ycallaer** [grafana/mimir#4009] [Help and support] [answered]
Hi all,
So I am pretty new to Mimir and I am currently a bit stuck. I have to write data to Mimir using python and found that you push metrics using the Mimir api, as documented [here](https://grafana.com/docs/mimir/latest/operators-guide/reference-http-api/#remote-write)

So I downloaded the proto definition , compiled it for python and constructed the message as per the proto requirements (all using dummy data so far).
However, when I do a post to the endpoint the HTTP return code is 400 and the error I get back is
```
received series has no metric name (err-mimir-missing-metric-name)
```
I did find a reference to this error in [this](https://grafana.com/docs/mimir/latest/operators-guide/mimir-runbooks/#err-mimir-missing-metric-name) document, but I can't figure out what I am missing in my code as all fields seem to be filled in.

The code -->
```
        label_pair = mimir_pb2.LabelPair(name=b"sampleLabel1", value=b"sampleValue1")
        sample_result = mimir_pb2.Sample(
            timestamp_ms=round(time.time() * 1000), value=1
        )
        exemplar_result = mimir_pb2.Exemplar(
            labels=[label_pair], value=1, timestamp_ms=round(time.time() * 1000)
        )
        timeseries_result = mimir_pb2.TimeSeries(
            labels=[label_pair], samples=[sample_result], exemplars=[exemplar_result]
        )

        metric_data = mimir_pb2.MetricMetadata(
            type=3, metric_family_name="metric-sample"
        )

        write_result = mimir_pb2.WriteRequest(
            timeseries=[timeseries_result],
            metadata=[metric_data],
            skip_label_name_validation=True,
        )
        write_result_binary = write_result.SerializeToString()
        print(write_result_binary)
        print("bin")
        data_compressed = snappy.compress(write_result_binary)

        response = requests.post(
            url="https://someurl:8080/api/v1/push",
            headers={
                "X-Prometheus-Remote-Write-Version": "0.1.0",
                "X-Mimir-SkipLabelNameValidation": "true",
                "Content-Encoding": "snappy",
                "Content-Type": "application/x-protobuf",
            },
            data=data_compressed,
            verify=False,
        )
        print(response.text)
        print(response.status_code)
```
I based myself a bit on the golang example for timeseries_test. Am I missing something in the code ?
Thanks for any insight you can provide

**Comment by Ycallaer** [comment]
So something I noticed in the proto definition is the following piece of code -->
```
message Metric {
  repeated LabelPair labels = 1 [(gogoproto.nullable) = false, (gogoproto.customtype) = "LabelAdapter"];
}

```
While it is defined though, it does not seem to be part of any proto definition, and hence not part of the WriteRequest. Maybe an issue ?

**Comment by pstibrany** [comment, marked as answer]
Hi. "Metric name" refers to a special label `__name__` that must be present in each series that Mimir ingests. Prometheus normally converts metric names into this label. For example series like:

```
# HELP node_cpu_seconds_total Seconds the CPUs spent in each mode.
# TYPE node_cpu_seconds_total counter
node_cpu_seconds_total{cpu="0",mode="idle"} 4282.67
node_cpu_seconds_total{cpu="0",mode="nice"} 0
node_cpu_seconds_total{cpu="0",mode="system"} 1009.48
node_cpu_seconds_total{cpu="0",mode="user"} 1471.18
```

All have metric name `node_cpu_seconds_total`. Prometheus internally stores these series as:
```
{__name__="node_cpu_seconds_total",cpu="0",mode="idle"}
{__name__="node_cpu_seconds_total",cpu="0",mode="nice"}
{__name__="node_cpu_seconds_total",cpu="0",mode="system"}
{__name__="node_cpu_seconds_total",cpu="0",mode="user"}
```

I don't see that label being set in your code.

**Reply by Ycallaer** [reply]
Hi, 
Thanks for the feedback, however I am unsure where this needs to be set ?
As per the documentation of mimir write api,  I used the proto file as defined [here](https://github.com/grafana/mimir/blob/main/pkg/mimirpb/mimir.proto) and set the fields to the same values as defined in timeseries_test.go

Since there is no __name__ definition there either , would it be possible to provide a small example ?

**Comment by Ycallaer** [comment]
Ok thanks,
I found what the issue was -->
`label_pair = mimir_pb2.LabelPair(name=b"__name__", value=b"sampleValue1")`
If I set this value than it post successful, thanks for the tip

**Reply by pstibrany** [reply]
Right. Note that single series can have multiple labels:

```
        label_pair1 = mimir_pb2.LabelPair(name=b"__name__", value=b"sampleValue1")
        label_pair2 = mimir_pb2.LabelPair(name=b"sampleLabel1", value=b"sampleValue1")

        timeseries_result = mimir_pb2.TimeSeries(
            labels=[label_pair1, label_pair2], ... # <-- here we add both labels (or even more, as needed).
        )
```
